### DISCUSSION TITLE
Math.random() creacion de intervalos

### DISCUSSION THREAD

**Discussion by UNE-Jose** [mmasias/24-25-PRG1#550] [General]
Respecto al ejercicio de Marco, vengo a explicar (a petición de don Manuel Masías) por qué en este caso no es necesario agregar el _**+1**_ que usualmente incluimos en los intervalos de números enteros.

Cuando usamos la expresión:

```
Math.random() * (MAXIMO - MINIMO + 1) + MINIMO
```

Esto genera un valor en el rango de $$[MINIMO, MAXIMO + 1) \in \mathbb{R}$$ debido a que estamos sumando 1 al intervalo. Esta suma es necesaria cuando queremos obtener un rango de números enteros ($$\mathbb{N} = \{1, 2, 3, 4, \dots\}$$).

Por otro lado, al usar la expresión:

```
(int)(Math.random() * (MAXIMO - MINIMO + 1) + MINIMO)
```

El uso de `(int)` hace que el resultado final sea un número entero en el intervalo $$[MINIMO, MAXIMO] \in \mathbb{N}$$, ya que la conversión a entero redondea hacia abajo el valor generado aleatoriamente.

Entonces, **para representar un intervalo continuo de números reales** se debería utilizar:

```
Math.random() * (MAXIMO - MINIMO) + MINIMO
```
Esta expresión generará un valor en el intervalo $$[MINIMO, MAXIMO) \in \mathbb{R}$$, es decir, un número real dentro del rango, sin necesidad de sumar 1.

**Comment by mmasias** [comment]
👏🏼👏🏼👏🏼👏🏼👏🏼👏🏼 Estupenda explicación, muchas gracias @UNE-Jose. Permíteme complementarla con un ejemplo concreto, recordando antes dos cosas: 

- Que `Math.random()` genera un número decimal aleatorio entre 0 (incluyéndolo) y 1 (sin incluirlo), es decir, puede ser 0.0 pero nunca llegará a ser 1.0
- Que _castear_ un valor de tipo `double` le quita su parte decimal, manteniendo la parte entera.

Ahora bien, poniendo MINIMO = 1 y MAXIMO = 3, tendríamos (si no usamos el +1): 

```java
Math.random() * (3 - 1) + 1
= Math.random() * 2 + 1
```
Lo cual  generará números entre:

- Mínimo: 1 (cuando Math.random() = 0.0)
- Máximo: 2.999... (cuando Math.random() se acerca a 0.999...)

Ahora bien, si usamos el +1:

```java
Math.random() * (3 - 1 + 1) + 1
= Math.random() * 3 + 1
```

Se generarán números entre:

- Mínimo: 1 (cuando Math.random() = 0.0)
- Máximo: 3.999... (cuando Math.random() se acerca a 0.999...)

Y cuando convirtamos estos números a enteros (_casteándolos_ a entero mediante `(int)`):

- Sin el +1: solo podemos obtener 1 y 2
- Con el +1: podemos obtener 1, 2 y 3

Por lo tanto, necesitamos el +1 para incluir el número MAXIMO en el rango de posibilidades cuando trabajamos con enteros. Sin el +1, el MAXIMO nunca sería alcanzable después de la conversión.
